### PR TITLE
 Added update service function chain unit test APIs to check virtual …

### PR DESCRIPTION
**Pull request by karimull**
…system order

### CODE REVIEW

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 1]
```diff
@@ -82,14 +84,67 @@ public void testDispatch_WhenVsIdsListUpdateValidationSuccessful_ValidationSucce
...
+		request.setVirtualSystemIds(vsIds);
+		Mockito.when(this.service.validator.validateAndLoad(request)).thenReturn(this.sfc);
+		// Act.
+		BaseJobResponse response = this.service.dispatch(request);
+		Assert.assertNotNull("The returned response should not be null.", response);
+	}
+	
+	@Test
+	public void testDispatch_sfcVirutalSystemIdsInOrder_ValidationSucceeds() throws Exception {
+
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: // Arrange.

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 2]
```diff
@@ -82,14 +84,67 @@ public void testDispatch_WhenVsIdsListUpdateValidationSuccessful_ValidationSucce
...
+		dto.setId(this.sfc.getId());
+		dto.setParentId(this.vc.getId());
+		request.setDto(dto);
+		request.setName(this.sfc.getName());
+		List<Long> vsIds = new ArrayList<Long>();
 		vsIds.add(this.vs1.getId());
 		vsIds.add(this.vs.getId());
 		request.setVirtualSystemIds(vsIds);
 		Mockito.when(this.service.validator.validateAndLoad(request)).thenReturn(this.sfc);
 		// Act.
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: add a break line

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 3]
```diff
@@ -82,14 +84,67 @@ public void testDispatch_WhenVsIdsListUpdateValidationSuccessful_ValidationSucce
...
+		//update sfc with virtual system ids list {2,1}
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: // Assert. The AAA comments really help to review and keep the longer tests like this organized.

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 4]
```diff
@@ -82,14 +84,67 @@ public void testDispatch_WhenVsIdsListUpdateValidationSuccessful_ValidationSucce
...
 		BaseJobResponse response = this.service.dispatch(request);
+		
+		//check Virtual system is as updated
+		List<Long> sfcVsIdList = this.em.find(ServiceFunctionChain.class, this.sfc.getId()).
+				getVirtualSystems().stream().map(vss -> vss.getId()).collect(Collectors.toList());
+		//read back sfc and cross check with local Virtual system list for order
+		Assert.assertEquals(vsIds, sfcVsIdList);
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: Please ensure to use a string message in all your asserts assertEquals("The list of virtual system ids is different than expected", ...). When the test fails this helps to quickly diagnose.  Also look at other asserts.

Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: Does this check the elements in the list?

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 5]
```diff
@@ -82,14 +84,67 @@ public void testDispatch_WhenVsIdsListUpdateValidationSuccessful_ValidationSucce
...
+		vsIds.add(this.vs1.getId());
+		request.setVirtualSystemIds(vsIds);
+		Mockito.when(this.service.validator.validateAndLoad(request)).thenReturn(this.sfc);
+		// Act.
+		BaseJobResponse response = this.service.dispatch(request);
+		Assert.assertNotNull("The returned response should not be null.", response);
+	}
+	
+	@Test
+	public void testDispatch_sfcVirutalSystemIdsInOrder_ValidationSucceeds() throws Exception {
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: I believe this is testing the service not the validation...then maybe ...validSFCVirtualSystemIds_SFCIsUpdated

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 6]
```diff
@@ -82,14 +84,67 @@ public void testDispatch_WhenVsIdsListUpdateValidationSuccessful_ValidationSucce
...
+		Mockito.when(this.service.validator.validateAndLoad(request)).thenReturn(this.sfc);
+		// Act.
+		//update sfc with virtual system ids list {2,1}
+		BaseJobResponse response = this.service.dispatch(request);
+		
+		//check Virtual system is as updated
+		List<Long> sfcVsIdList = this.em.find(ServiceFunctionChain.class, this.sfc.getId()).
+				getVirtualSystems().stream().map(vss -> vss.getId()).collect(Collectors.toList());
+		
+		//change the order to compare
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: why is this needed?

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 7]
```diff
@@ -82,14 +84,67 @@ public void testDispatch_WhenVsIdsListUpdateValidationSuccessful_ValidationSucce
...
 		// Act.
+		//update sfc with virtual system ids list {2,1}
 		BaseJobResponse response = this.service.dispatch(request);
+		
+		//check Virtual system is as updated
+		List<Long> sfcVsIdList = this.em.find(ServiceFunctionChain.class, this.sfc.getId()).
+				getVirtualSystems().stream().map(vss -> vss.getId()).collect(Collectors.toList());
+		//read back sfc and cross check with local Virtual system list for order
+		Assert.assertEquals(vsIds, sfcVsIdList);
 		Assert.assertNotNull("The returned response should not be null.", response);
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: Can you also check something about the content of the response? Maybe the expected id?

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 8]
```diff
@@ -82,14 +84,67 @@ public void testDispatch_WhenVsIdsListUpdateValidationSuccessful_ValidationSucce
...
 	}
 	
-	//TODO : karimull  Add more test cases in looking for updated in order. Need some ramp up here
+	@Test
+	public void testDispatch_sfcVirutalSystemIdsNotInOrder_ValidationFails() throws Exception {
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: This name does not align with the test code. The validator is not mocked to fail and the sfc seems to be updated. So in essence this tests looks like the previous one.

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java (hunk 4 above)
Comment (karimull) [reply osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: Yes it is checking all the list elements.

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java (hunk 8 above)
Comment (karimull) [reply osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: Agreed , it is kind of similar to previous one, but thought of checking in reverse also. 
May not needed ..taking it off.

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java (hunk 6 above)
Comment (karimull) [reply osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: removed the test

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java (hunk 7 above)
Comment (karimull) [reply osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: added check

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 9]
```diff
@@ -60,91 +60,77 @@ public void testDispatch_WithNullRequest_ThrowsNullPointerException() throws Exc
 	@Test
 	public void testDispatch_WithNonEmptyListOfVsIdsUpdateValidationSuccessful_ValidationSucceeds() throws Exception {
 		
+		//Arrange
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: picky `//Arrange` -> `// Arrange.` Observe the space and period. Also check other places.

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 10]
```diff
@@ -60,91 +60,77 @@ public void testDispatch_WithNullRequest_ThrowsNullPointerException() throws Exc
 	@Test
 	public void testDispatch_WithNonEmptyListOfVsIdsUpdateValidationSuccessful_ValidationSucceeds() throws Exception {
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: The name is not accurate. You are not testing that the validation succeeds, the validator is mocked it should have its own tests. You are testing that **when** the validation succeeds the SFC is updated. This names gives the impression you are testing the validation which is being mocked. Maybe something like testDispatch_WithNonEmpty..._SFCIsUpdated

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 11]
```diff
@@ -60,91 +60,77 @@ public void testDispatch_WithNullRequest_ThrowsNullPointerException() throws Exc
...
 		vsIds.add(this.vs.getId());
 		vsIds.add(this.vs1.getId());
 		request.setVirtualSystemIds(vsIds);
 		Mockito.when(this.service.validator.validateAndLoad(request)).thenReturn(this.sfc);
+		
 		// Act.
 		BaseJobResponse response = this.service.dispatch(request);
+		
+		//Assert
 		Assert.assertNotNull("The returned response should not be null.", response);
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: Also please validate what should be in the response.

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 12]
```diff
@@ -60,91 +60,77 @@ public void testDispatch_WithNullRequest_ThrowsNullPointerException() throws Exc
...
-		//update sfc with virtual system ids list {2,1}
 		BaseJobResponse response = this.service.dispatch(request);
 		
-		//check Virtual system is as updated
-		List<Long> sfcVsIdList = this.em.find(ServiceFunctionChain.class, this.sfc.getId()).
-				getVirtualSystems().stream().map(vss -> vss.getId()).collect(Collectors.toList());
-		//read back sfc and cross check with local Virtual system list for order
-		Assert.assertEquals(vsIds, sfcVsIdList);
+		//Assert
 		Assert.assertNotNull("The returned response should not be null.", response);
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: Again, assert what should be in the response. Also how do we know that the service is calling the db with the correct info? If there is a regression on that code this is not going to catch it.

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 13]
```diff
@@ -60,91 +60,77 @@ public void testDispatch_WithNullRequest_ThrowsNullPointerException() throws Exc
...
-		
-		//change the order to compare
-		vsIds.clear();
-		vsIds.add(this.vs.getId());
-		vsIds.add(this.vs1.getId());
 		//read back sfc and cross check with local Virtual system list for order
-		Assert.assertNotEquals(vsIds, sfcVsIdList);
+		
+		//Assert
+		Assert.assertEquals("The list of virtual system ids is different than expected", vsIds, sfcVsIdList);
```
Comment (emanoelxavier) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: Is this test the same as the previous one but with additional validation? If so you should have just a single test case containing all the validation/asserts that are applicable. You would have a new test case only if the inputs, action or mocked state changes but you can pack as much validation as possible on the same test method/case

Comment (karimull) [reply osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: Added three test cases with Empty list of Virtual systems, a updated virtual systems and last one is Validation of the order. Yes the last two are similar, but do not want to stub into everything as I change the input list to check back the action.

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java (hunk 11 above)
Comment (karimull) [reply osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: added response id check

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java (hunk 12 above)
Comment (karimull) [reply osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: added response id check

**Comment by karimull** [comment]
Removed one redundant test case and updated names

File: osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java [hunk 14]
```diff
@@ -56,40 +58,59 @@ public void testDispatch_WithNullRequest_ThrowsNullPointerException() throws Exc
...
-	
+
+
 	@Test
-	public void testDispatch_WhenVsIdsListUpdateValidationSuccessful_ValidationSucceeds() throws Exception {
-		
+	public void testDispatch_WhenVirtualSystemListNotEmpty_SfcIsUpdatedAndListInOrder() throws Exception {
+
+		// Arrange
 		BaseDto dto = new BaseDto();
```
Comment (arvindn05) [inline osc-server/src/test/java/org/osc/core/broker/service/servicefunctionchain/UpdateServiceFunctionChainServiceTest.java]: since you are doing the same in multiple places can you just do this in testInitialize?
